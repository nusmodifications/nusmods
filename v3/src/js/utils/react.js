// @flow

import type { Node } from 'react';

// Define some useful Unicode characters as constants
export const NBSP = '\u00a0';
export const BULLET = ' â€¢ ';

/**
 * Replace substring matching the provided regex with React nodes. This is
 * basically the React version of replacing bits of strings with HTML tags,
 * except it is safer as the returned result is a React Node, so there's
 * no need to set innerHTML. This is useful for eg. highlighting search results
 * with <mark>
 *
 * @param str
 * @param regex
 * @param replacement
 * @returns {Node}
 */
export function replaceWithNode(
  str: string,
  regex: RegExp,
  replacement: (match: string, index: number) => Node,
): Node {
  const parts = str.split(regex);

  // We want to ensure the resulting array always have the matches at even position
  // eg. ['Some text ', 'CS1010S', ' more text ', 'CS3216', 'more text']
  // This allows us to replace the even position elements with components.
  // However, if the string starts with a match, then the first element will be a match
  // so we add in an empty string to pad matches to even positions
  if (parts.length && regex.test(parts[0])) parts.unshift('');

  return parts.map((part, i) => {
    if (i % 2 === 0) return part;
    return replacement(part, (i - 1) / 2);
  });
}

/**
 * Replaces all spaces in a string with U+00A0 non-breaking spaces. This
 * avoids the browser inserting a page break between text that are part
 * of the same logical unit in the UI, such as 'Semester 1' or 'Week 13'.
 * Avoid using this for content.
 *
 * @param {string} text
 * @returns {string}
 */
export function noBreak(text: string): string {
  return text.replace(/ /g, NBSP);
}
