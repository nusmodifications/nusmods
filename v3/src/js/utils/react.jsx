// @flow

import type { Node } from 'react';
import React from 'react';
import { escapeRegExp, castArray, entries } from 'lodash';

// Define some useful Unicode characters as constants
export const NBSP = '\u00a0';
export const BULLET = ' â€¢ ';

// Define media breakpoints
type Breakpoint = 'xs' | 'sm' | 'md' | 'lg' | 'xl';
const breakpoints: { [Breakpoint]: number } = {
  xs: 0,
  sm: 576,
  md: 768,
  lg: 992,
  xl: 1200,
};

function nextBreakpoint(size: Breakpoint): ?Breakpoint {
  const breakpointEntries = entries(breakpoints);
  const nextBreakpointIndex = breakpointEntries.findIndex(([breakpoint]) => breakpoint === size) + 1;
  if (nextBreakpointIndex >= breakpointEntries.length) return null;
  return breakpointEntries[nextBreakpointIndex][1];
}

export function breakpointDown(size: Breakpoint): MediaQueryList {
  const nextSize = nextBreakpoint(size);
  if (nextSize == null) return window.matchMedia('all');
  return window.matchMedia(`(max-width: ${breakpoints[nextSize] - 1}px )`);
}

export function breakpointUp(size: Breakpoint): MediaQueryList {
  return window.matchMedia(`(min-width: ${breakpoints[size]}px)`);
}

/**
 * Replace substring matching the provided regex with React nodes. This is
 * basically the React version of replacing bits of strings with HTML tags,
 * except it is safer as the returned result is a React Node, so there's
 * no need to set innerHTML. This is useful for eg. highlighting search results
 * with <mark>
 *
 * @param str
 * @param regex
 * @param replacement
 * @returns {Node}
 */
export function replaceWithNode(
  str: string,
  regex: RegExp,
  replacement: (match: string, index: number) => Node,
): Node {
  const parts = str.split(regex);

  // We want to ensure the resulting array always have the matches at even position
  // eg. ['Some text ', 'CS1010S', ' more text ', 'CS3216', 'more text']
  // This allows us to replace the even position elements with components.
  // However, if the string starts with a match, then the first element will be a match
  // so we add in an empty string to pad matches to even positions
  if (parts.length && regex.test(parts[0])) parts.unshift('');

  return parts.map((part, i) => {
    if (i % 2 === 0) return part;
    return replacement(part, (i - 1) / 2);
  });
}

export function highlight(str: string, search: string | string[], Tag: string = 'mark'): Node {
  const terms = castArray(search).filter(Boolean);
  if (!terms.length) return str;
  const regex = new RegExp(`(${terms.map(escapeRegExp).join('|')})`, 'ig');
  return replaceWithNode(str, regex, (match, i) => <Tag key={i}>{match}</Tag>);
}

/**
 * Replaces all spaces in a string with U+00A0 non-breaking spaces. This
 * avoids the browser inserting a page break between text that are part
 * of the same logical unit in the UI, such as 'Semester 1' or 'Week 13'.
 * Avoid using this for content.
 *
 * @param {string} text
 * @returns {string}
 */
export function noBreak(text: string): string {
  return text.replace(/ /g, NBSP);
}

export function defer(task: () => any) {
  window.requestAnimationFrame(() => {
    window.requestAnimationFrame(task);
  });
}
