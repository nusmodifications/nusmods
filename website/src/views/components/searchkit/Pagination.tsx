import React from 'react';
import { PaginationAccessor, SearchkitComponent, SearchkitComponentProps } from 'searchkit';
import { clamp, get } from 'lodash';

export interface PagerProps {
  selectedPage: number;
  totalNumPages: number;
  onGoToFirst: () => void;
  onGoToPrevious: () => void;
  onGoToPage: (page: number) => void;
  onGoToNext: () => void;
  onGoToLast: () => void;
}

export interface PaginationProps extends SearchkitComponentProps {
  pagerComponent: React.ElementType<PagerProps>;
}

export const FIRST_PAGE_INDEX = 1;

// Custom equivalent of Searchkit's Pagination component
// eslint-disable-next-line @typescript-eslint/ban-types
export default class Pagination extends SearchkitComponent<PaginationProps, {}> {
  paginationAccessor() {
    return this.accessor as PaginationAccessor;
  }

  // eslint-disable-next-line class-methods-use-this
  defineAccessor() {
    return new PaginationAccessor('p');
  }

  getCurrentPage() {
    return Number(this.paginationAccessor().state.getValue()) || FIRST_PAGE_INDEX;
  }

  getTotalPages() {
    return Math.ceil(this.getHitsCount() / get(this.getQuery(), 'query.size', 10));
  }

  onGoToFirst = () => this.setPage(FIRST_PAGE_INDEX);

  onGoToPrevious = () => this.setPage(this.getCurrentPage() - 1);

  onGoToPage = (page: number) => this.setPage(page);

  onGoToNext = () => this.setPage(this.getCurrentPage() + 1);

  onGoToLast = () => this.setPage(this.getTotalPages());

  setPage(requestedPage: number) {
    const page = clamp(requestedPage, FIRST_PAGE_INDEX, this.getTotalPages());
    if (page === this.getCurrentPage()) {
      return; // Same page, no need to rerun query
    }
    this.paginationAccessor().state = this.paginationAccessor().state.setValue(page);
    this.searchkit.performSearch();
  }

  render() {
    if (!this.paginationAccessor()) return null;
    if (!this.hasHits()) return null;

    const { pagerComponent: Pager } = this.props;
    return (
      <Pager
        selectedPage={this.getCurrentPage()}
        totalNumPages={this.getTotalPages()}
        onGoToFirst={this.onGoToFirst}
        onGoToPrevious={this.onGoToPrevious}
        onGoToPage={this.onGoToPage}
        onGoToNext={this.onGoToNext}
        onGoToLast={this.onGoToLast}
      />
    );
  }
}
