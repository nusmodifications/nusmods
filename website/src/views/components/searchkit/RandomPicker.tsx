import {
  PageSizeAccessor,
  PaginationAccessor,
  SearchkitComponent,
  SearchkitComponentProps,
} from 'searchkit';
import { ModuleCode } from 'types/modules';

export type RandomPickerProps = {
  getRandomModuleCode: () => Promise<ModuleCode>;
};

interface SearchkitRandomPickerProps extends SearchkitComponentProps {
  buttonComponent: React.ElementType<RandomPickerProps>;
}

type State = Record<string, never>;

export default class RandomPicker extends SearchkitComponent<SearchkitRandomPickerProps, State> {
  paginationAccessor() {
    return this.accessor as PaginationAccessor;
  }

  // eslint-disable-next-line class-methods-use-this
  override defineAccessor() {
    return new PaginationAccessor('p');
  }

  getRandomModuleCode = async (): Promise<ModuleCode> => {
    const sizeAccessor = this.searchkit.getAccessorByType(PageSizeAccessor) as PageSizeAccessor;
    const totalPages = Math.ceil(this.searchkit.getHitsCount() / sizeAccessor.getSize());
    const randomPage = Math.floor(Math.random() * totalPages);

    this.paginationAccessor().state = this.paginationAccessor().state.setValue(randomPage);
    const { hits } = (await this.searchkit.performSearch()).results.hits;
    const randomHit = hits[Math.floor(Math.random() * hits.length)];
    /* eslint-disable no-underscore-dangle */
    return randomHit._source.moduleCode;
  };

  override render() {
    const { buttonComponent: Button } = this.props;
    return <Button getRandomModuleCode={this.getRandomModuleCode} />;
  }
}
