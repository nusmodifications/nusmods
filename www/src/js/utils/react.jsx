// @flow

import type { Node, ComponentType } from 'react';
import React from 'react';
import { escapeRegExp, castArray } from 'lodash';

// Define some useful Unicode characters as constants
export const NBSP = '\u00a0';
export const BULLET = ' â€¢ ';

/**
 * Replace substring matching the provided regex with React nodes. This is
 * basically the React version of replacing bits of strings with HTML tags,
 * except it is safer as the returned result is a React Node, so there's
 * no need to set innerHTML. This is useful for eg. highlighting search results
 * with <mark>
 *
 * @param str
 * @param regex
 * @param replacement
 * @returns {Node}
 */
export function replaceWithNode(
  str: string,
  regex: RegExp,
  replacement: (match: string, index: number) => Node,
): Node {
  const parts = str.split(regex);

  // We want to ensure the resulting array always have the matches at even position
  // eg. ['Some text ', 'CS1010S', ' more text ', 'CS3216', 'more text']
  // This allows us to replace the even position elements with components.
  // However, if the string starts with a match, then the first element will be a match
  // so we add in an empty string to pad matches to even positions
  if (parts.length && regex.test(parts[0])) parts.unshift('');

  return parts.map((part, i) => {
    if (i % 2 === 0) return part;
    return replacement(part, (i - 1) / 2);
  });
}

export function highlight(str: string, search: string | string[], Tag: string = 'mark'): Node {
  const terms = castArray(search).filter(Boolean);
  if (!terms.length) return str;
  const regex = new RegExp(`(${terms.map(escapeRegExp).join('|')})`, 'ig');
  return replaceWithNode(str, regex, (match, i) => <Tag key={i}>{match}</Tag>);
}

/**
 * Replaces all spaces in a string with U+00A0 non-breaking spaces. This
 * avoids the browser inserting a page break between text that are part
 * of the same logical unit in the UI, such as 'Semester 1' or 'Week 13'.
 * Avoid using this for content.
 *
 * @param {string} text
 * @returns {string}
 */
export function noBreak(text: string): string {
  return text.replace(/ /g, NBSP);
}

export function defer(task: () => any) {
  window.requestAnimationFrame(() => {
    window.requestAnimationFrame(task);
  });
}

export function wrapComponentName(Component: ComponentType, wrapper: string): string {
  return `${wrapper}(${Component.displayName || Component.name || 'Component'})`;
}

/**
 * Small utility function to scroll to an element with ID matching the URL hash if
 * both are present.
 *
 * This mimics traditional webpage behavior and should be used in componentDidMount() when
 * the component is not loaded on initial page load (ie. the element is not in the DOM when
 * the page is initially loaded), but has content that can be linked to via hashes.
 */
export function scrollToHash() {
  const hash = window.location.hash;
  if (hash) {
    const ele = document.getElementById(hash.slice(1)); // Hash string contains the '#' character
    if (ele) {
      ele.scrollIntoView(true);
    }
  }
}

/**
 * Utility class that encapsulates an auto-incrementing counter. Useful for
 * keeping track of Downshift item indices
 */
export class Counter {
  count = -1;

  index() {
    this.count += 1;
    return this.count;
  }

  matches(index: number) {
    this.count += 1;
    return this.count === index;
  }
}
