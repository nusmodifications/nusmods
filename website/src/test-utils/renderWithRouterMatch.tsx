import { render } from '@testing-library/react';
import type { ReactNode } from 'react';
import { Route, Router } from 'react-router-dom';
import createHistory from './createHistory';

/**
 * `render` `children` in a `Router` and `Route` so that `children` have
 * populated route matches when using React Router.
 *
 * Inspiration: https://spectrum.chat/testing-library/help-react/attempting-to-test-react-router-match~b0550426-f54a-4b76-b402-c7b32204b55e?m=MTU2OTM1MzY4NjUwNw==
 */
export default function renderWithRouterMatch(
  children: ReactNode,
  {
    path = '/',
    location = '/',
  }: {
    path?: string;
    location?: Parameters<typeof createHistory>[0];
  },
) {
  const { history } = createHistory(location);
  const view = render(
    <Router history={history}>
      <Route path={path}>{children}</Route>
    </Router>,
  );
  return {
    history,
    view,
  };
}
