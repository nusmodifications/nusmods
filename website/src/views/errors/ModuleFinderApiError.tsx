import React from 'react';
import { SearchkitManager } from 'searchkit';
import ApiError from './ApiError';

type ModuleFinderApiErrorProps = {
  searchkit: SearchkitManager;
};

const ModuleFinderApiError: React.FC<ModuleFinderApiErrorProps> = ({ searchkit }) => (
  <ApiError dataName="course information" retry={() => searchkit.reloadSearch()} />
);

export default ModuleFinderApiError;
