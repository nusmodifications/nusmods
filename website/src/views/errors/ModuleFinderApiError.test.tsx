import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { SearchkitManager } from 'searchkit';

import renderWithRouterMatch from 'test-utils/renderWithRouterMatch';
import ModuleFinderApiError from './ModuleFinderApiError';

jest.mock('views/components/RandomKawaii');

const mockedUrl = 'https://someEsUrl.com';
const CANONICAL = '/modules?q=GER1000&sem[0]=1&sem[1]=2&sem[2]=3&sem[3]=4';

const searchkit = new SearchkitManager(mockedUrl, {
  searchUrlPath: '_search?track_total_hits=true',
});

describe(ModuleFinderApiError, () => {
  test('should render', async () => {
    const { container } = render(<ModuleFinderApiError searchkit={searchkit} />);
    expect(container).toMatchSnapshot();
  });

  test('should render with url', async () => {
    const { view } = renderWithRouterMatch(<ModuleFinderApiError searchkit={searchkit} />, {
      path: '/modules',
      location: CANONICAL,
    });
    const { container } = view;
    expect(container).toMatchSnapshot();
  });

  // TODO: Tech Debt - Check if this test can correctly perform an assertion on `history`
  test('should have unchanged url when search is retried', async () => {
    const { history, view } = renderWithRouterMatch(
      <ModuleFinderApiError searchkit={searchkit} />,
      {
        path: '/modules',
        location: CANONICAL,
      },
    );
    expect(history.location.pathname).toBe('/modules');
    expect(history.location.search).toBe('?q=GER1000&sem[0]=1&sem[1]=2&sem[2]=3&sem[3]=4');

    const { getByRole } = view;
    searchkit.history = {
      push: jest.fn(),
    };
    const tryAgainButton = getByRole('button');
    await userEvent.click(tryAgainButton);
    expect(searchkit.history.push).toHaveBeenCalled();
    expect(history.location.pathname).toBe('/modules');
    expect(history.location.search).toBe('?q=GER1000&sem[0]=1&sem[1]=2&sem[2]=3&sem[3]=4');
  });
});
