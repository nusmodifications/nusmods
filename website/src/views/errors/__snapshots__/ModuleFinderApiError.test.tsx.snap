// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModuleFinderApiError should render 1`] = `
<div>
  <div>
    <div
      class="container"
    >
      <div
        class="header"
      >
        <div
          data-testid="RandomKawaii component"
        />
      </div>
      <h1
        class="h3 header"
      >
        <span
          class="expr"
        >
          Oh no...
        </span>
         
        We can't load the course information
      </h1>
      <p>
        This could be because your device is offline or NUSMods is down :(
      </p>
      <strong>
        Course search might be having issues at the moment. ðŸ˜Ÿ
      </strong>
      <p>
        If it isn't working, please try the module search
         
        on the top right corner of the page instead.
      </p>
      <div>
        <button
          class="btn btn-primary btn-lg"
          type="button"
        >
          Click to try again
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ModuleFinderApiError should render with url 1`] = `
<div>
  <div>
    <div
      class="container"
    >
      <div
        class="header"
      >
        <div
          data-testid="RandomKawaii component"
        />
      </div>
      <h1
        class="h3 header"
      >
        <span
          class="expr"
        >
          Oh no...
        </span>
         
        We can't load the course information
      </h1>
      <p>
        This could be because your device is offline or NUSMods is down :(
      </p>
      <strong>
        Course search might be having issues at the moment. ðŸ˜Ÿ
      </strong>
      <p>
        If it isn't working, please try the module search
         
        on the top right corner of the page instead.
      </p>
      <div>
        <button
          class="btn btn-primary btn-lg"
          type="button"
        >
          Click to try again
        </button>
      </div>
    </div>
  </div>
</div>
`;
